Kernel

virgile-baudrot opened the pull request:
# Description

Additional kernels: Geometric, Gaussian and Bivariate Student 
Formulation founds in Nathan 2012, "Dispersal kernels: review" from the book Dispersal Ecology and Evolution (2012) and Walker et al. (2019) "A Spatio-Temporal Exposure-Hazard Model for Assessing Biological Risk and Impact"

❓  How (where) to cite the sources?

## Changes (choose one only, with x or click after creating PR):

- [x] New formulation/rule
- [ ] Documentation
- [ ] Bug fix for existing rules (non-breaking syntax)
- [ ] Minor change to existing rules (no breaks to old syntax)
- [ ] Major change to existing rules (breaking syntax, needs a breaking version bump)

# Checklist:

- [x] My code follows the [BlueStyle](https://github.com/invenia/BlueStyle) style guide
- [x] My code is commented with explanations of anything hard to understand
- [x] There are no unnecessary or unrelated code changes included in this PR
- [x] There are no dependencies added: if so, please consider if they are necessary and explain why they are required

### For bugfixes

- [ ] My PR addresses a single bug/connected group of bugs
- [ ] I have included a test that will ensure the bug does not reoccur

### For new `Rule`s

- [x] My PR is limited to a clear theme/conceptual unit (otherwise make multiple PRs that can be discussed separately).
- [x] New rules have tests in a file of the same name as the rule file, e.g. `/test/newrulename.jl`
- [ ] I have cited the source of the new formulation/s, if applicable
- [x] I have added a documentation string to formulation struct/s, and added an entry in the docs

virgile-baudrot committed:
add mortality functions

virgile-baudrot committed:
export mortality functions  + test

virgile-baudrot committed:
add Mortality functions to index doc

virgile-baudrot committed:
add additional kernels: Geometric, Gaussian and Student

virgile-baudrot committed:
add additional kernels: Geometric, Gaussian and Student

virgile-baudrot committed:
add kernel in index of docs

virgile-baudrot committed:
add kernel in index of docs

virgile-baudrot committed:
fix BivariateStudentkernel kernels boundaries

virgile-baudrot committed:
fix BivariateStudentkernel kernels boundaries

virgile-baudrot committed:
add test for new kernels

virgile-baudrot committed:
add test for new kernels

jamesmaino reviewed (approved):
Great! Perhaps we should cite in the documentation string for each function?

rafaqz commented on src/Dispersal.jl:
```diff
@@ -20,7 +20,7 @@ using ModelParameters: params
 import DynamicGrids: applyrule, applyrule!, precalcrule, kernel
 
-export KernelFormulation, ExponentialKernel
+export KernelFormulation, ExponentialKernel, GeometricKernel, GaussianKernel, BivariateStudentKernel
```
Do you think we should remove `Kernel` from the end of all of these? I would be cleaner for scripting, and its a bit confusing that `DisperselKernel` also ends in `Kernel`.

virgile-baudrot replied:
We have other rules using word like Exponential in `ExponentialGrowth` for instance, so keeping `Kernel` key word is maybe valuable.

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
+@Base.kwdef struct GeometricKernel{P} <: KernelFormulation
+    α::P = Param(1.0, bounds=(-1000.0, 1000.0))
+end
+(f::GeometricKernel)(distance) = (1+distance)^f.α * ((f.α+1)*(f.α+2))/(2*π)
```
spaces around math, except `^`

And `d`  for `distance`?

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
     λ::P = Param(1.0, bounds=(0.0, 2.0))
 end
 (f::ExponentialKernel)(distance) = exp(-distance / f.λ)
```
Should we use `d` everywhere for distance, seeing it is used below?

virgile-baudrot replied:
Yes `d` is fine. `distance` is quite big when formulation is long. There is also `r` as radius which is often used (in Nathan 2012).

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
+@Base.kwdef struct GaussianKernel{P} <: KernelFormulation
+    α::P = Param(1.0, bounds=(0.0, 1000.0))
+end
+(f::GaussianKernel)(d) = 1/(π*f.α^2)*exp(-d^2/f.α^2)
```
spaces around math except `^` and singular `-`

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
+    c1::P4 = Param(0.5, bounds=(0.0, 10.0))
+    c2::P5 = Param(0.5, bounds=(0.0, 2*π))
+end
+(f::BivariateStudentKernel)(d) = (f.β-1)/(π*f.α^2)*(1+d^2/f.α^2)*exp(f.c1 * cos(f.θ-f.c2)) 
```
spaces around math

rafaqz commented on test/kernels.jl:
```diff
@@ -165,3 +166,29 @@ end
+    @testset "ExponentialKernel" begin
+        hood = DispersalKernel(; formulation=ExponentialKernel(0.5))
+        @test DynamicGrids.radius(hood) == 1
+        @test sum(DynamicGrids.kernel(hood)) ≈ 1 atol=10^(-10)
```
`1e-10` for `10^(-10)`, and the rest below

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -32,3 +32,69 @@ where λ is a shape parameter.
+    BivariateStudentKernel(α, β, θ, c1, c2)
+
+Probability density function of distance ``d``.
+The Bivariate Student kernel has a power-law decrease and is anisotropic.
```
Maybe explain it in slightly more lay terms as well, maybe give an example use-case for a directional kernel?

virgile-baudrot replied:
ok. and the parameterisation I used is isotropic 👎

rafaqz reviewed (changes requested):
This is great. Thanks for writing these up.

I'm wondering if we should drop the `Kernel` from the end, so you just do `formulation=Gaussian(a)`. It would be nicer for scripts.

I would put the citation in every doc string.

virgile-baudrot commented:
>  drop the `Kernel` from the end, so you just do `formulation=Gaussian(a)`

I thought about that also. For Gaussian and Geometric it's working. But we have `ExponentialKernel`, `ExponentialGrowth` and `ExponentialMortality`... what makes many exponential, and why calling one Exponential and not the other

rafaqz commented:
That's a good point. `Exponential` needs to be clarified. `Geometric` is also the same name as `Distributions.Geometric` - so maybe leaving `Kernel` on the end is better.

virgile-baudrot committed:
remove fonction with additional Grid and test `Grid(:X)`

virgile-baudrot committed:
clean documentations

virgile-baudrot committed:
remove extra whitespace

rafaqz committed:
propose_docs_example

virgile-baudrot committed:
Merge remote-tracking branch 'origin/kernel' into kernel

rafaqz commented:
Any movement on this one virgile? Youve probably been busy. But great to have it in the package for submission

virgile-baudrot committed:
change 'distance' key to 'd' for consistancy

virgile-baudrot committed:
remove Bivariate Student, Add Weibull

virgile-baudrot committed:
pdate documentation and export

virgile-baudrot committed:
update doc export

virgile-baudrot commented:
I made some updates on type of wirtting equations.
Also I added Weibull kernel which is very use for pollen dispersal
I removed BivarianteStudent, I'm not fully sure about the mathematical formulation in Nathan et al.

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
+
+where α is a shape parameter.
+"""
+@Base.kwdef struct GeometricKernel{P} <: KernelFormulation
```
This `@` should be before `kwdef` :)

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
+
+where α is a positive parameter.
+"""
+@Base.kwdef struct GaussianKernel{P} <: KernelFormulation
```
This one too

rafaqz commented on src/kernel/formulations.jl:
```diff
@@ -31,4 +31,63 @@ where λ is a shape parameter.
+
+where α and β are positive parameters.
+"""
+@Base.kwdef struct WeibullKernel{A,B} <: KernelFormulation
```
And this

rafaqz commented:
Also your commits are a bit messed up with some of my commits somehow... we will have to rebase this down into a single commit here (which is easy to do), or otherwise if you want to rebase against master and then rebase you commits down to a few representative ones?

When working on a PR you should always `rebase` it against master and other branches, instead of `merge`. `rebase` just rewrites all your commits at the end.

virgile-baudrot committed:
update kernel formulations

virgile-baudrot commented:
arrff, sorry for that. I made a change on my forks forgetting it follows the PR here

I'm doing a new branch base on the master for kernel... come in few minutes

virgile-baudrot commented:
@rafaqz how do I PR here with a this new branch [kernel_PR](https://github.com/virgile-baudrot/Dispersal.jl/commit/6197a0c214edf3547f1f8561e5e3967d96690cde) which is on my fork? Should I start a new PR and close this one?

rafaqz commented:
You can push another branch to this one with `--force`

virgile-baudrot commented:
move to #98
